src/domain/task-system/domain/src/entities/logical_plan.rs

```diff
@@ -74,6 +77,15 @@ pub struct HardCompactionDataset {
 
 ////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
 
+/// A task to perform the resetting of a dataset
+#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
+pub struct ResetDataset {
+    pub dataset_id: DatasetID,
+    pub new_head_hash: Multihash,
```

Comment:
Reset is a destructive operation that can lead to data loss.

We better be 100% sure that we are only loosing data that the user wanted to loose and **not one bit more**.

Imagine a situation where:
- User creates a reset flow
- For whatever reason the flow system is stuck for a significant amount of time
- During this time user's IoT device continues to write data and get HTTP 200 responses that data was durably stored
- We fix the flow system and it happily resets the dataset and all the new data that arrived

We should be making **optimistic concurrency control** a part of every destructive flow.

So I think we should add `old_head_hash` parameter to the task, so that it aborts if it sees that dataset has changed since the decision was made to reset the dataset.

The handling of this parameter likely belongs all the way down in `ResetService`, so that in future we could make it fully atomic when `head` pointer is stored in a database.

And I think this parameter should originate from the API caller - only this way we can be 100% sure that we never reset something that user cared about.

Reply:
Yeah. It is actually a great improvement. Thanks